clock/README.md

```diff
@@ -5,6 +5,362 @@ A Go time testing library for writing deterministic unit tests
+```
+
+### Traps
+
+A trap allows you to match specific calls into the library while mocking, block their return,
+inspect their arguments, then release them to allow them to return. They help you write
+deterministic unit tests even when the code under test executes asynchronously from the test.
+
+You set your traps prior to executing code under test, and then wait for them to be triggered.
+
+```go
+func TestTrap(t *testing.T) {
+    ctx, cancel := context.WithTimeout(10*time.Second)
+    defer cancel()
+    mClock := quartz.NewMock(t)
+    trap := mClock.Trap().AfterFunc()
+    defer trap.Close() // stop trapping AfterFunc calls
+
+    count := 0
+    go mClock.AfterFunc(time.Hour, func(){
+        count++
+    })
+    call := trap.MustWait(ctx)
+    call.Release()
+    if call.Duration != time.Hour {
+        t.Fatal("wrong duration")
+    }
+
+    // Now that the async call to AfterFunc has occurred, we can advance the clock to trigger it
+    mClock.Advance(time.Hour).MustWait(ctx)
+    if count != 1 {
+        t.Fatal("wrong count")
+    }
+}
+```
+
+In this test, the trap serves 2 purposes. Firstly, it allows us to capture and assert the duration
+passed to the `AfterFunc` call. Secondly, it prevents a race between setting the timer and advancing
+it. Since these things happen on different goroutines, if `Advance()` completes before
+`AfterFunc()` is called, then the timer never pops in this test.
+
+Any untrapped calls immediately complete using the current time, and calling `Close()` on a trap
+causes the mock clock to stop trapping those calls.
+
+You may also `Advance()` the clock between trapping a call and releasing it. The call uses the
+current (mocked) time at the moment it is released.
+
+```go
+func TestTrap2(t *testing.T) {
+    ctx, cancel := context.WithTimeout(10*time.Second)
+    defer cancel()
+    mClock := quartz.NewMock(t)
+    trap := mClock.Trap().Since()
+    defer trap.Close() // stop trapping AfterFunc calls
+
+    var logs []string
+    done := make(chan struct{})
+    go func(clk quartz.Clock){
+        defer close(done)
+        start := clk.Now()
+        phase1()
+        p1end := clk.Now()
+        logs = append(fmt.Sprintf("Phase 1 took %s", p1end.Sub(start).String()))
+        phase2()
+        p2end := clk.Now()
+        logs = append(fmt.Sprintf("Phase 2 took %s", p2end.Sub(p1end).String()))
+    }(mClock)
+
+    // start
+    trap.MustWait(ctx).Release()
+    // phase 1
+    call := trap.MustWait(ctx)
+    mClock.Advance(3*time.Second).MustWait(ctx)
+    call.Release()
+    // phase 2
+    call = trap.MustWait(ctx)
+    mClock.Advance(5*time.Second).MustWait(ctx)
+    call.Release()
+
+    <-done
+    // Now logs contains []string{"Phase 1 took 3s", "Phase 2 took 5s"}
+}
+```
+
+### Tags
+
+When multiple goroutines in the code under test call into the Clock, you can use `tags` to
+distinguish them in your traps.
+
+```go
+trap := mClock.Trap.Now("foo") // traps any calls that contain "foo"
```

Comment:
I find tag systems have a persistent cognitive overhead as I have to remember the subtle differences between implementations. E.g.:
* do tags match exactly or does a subset suffice
 * if I trap "foo" and call w/ "foo" "bar" does it match?
 * if I trap "foo" "bar" and call w/ "bar" does it match?
* is matching case sensitive?

And then, there's a long-term burden of keeping the tag names consistent.

Instead, I wonder if we should promote accepting multiple independent clock instances.

Reply:
> I wonder if we should promote accepting multiple independent clock instances.

I think that would significantly increase the friction of adopting this library. In simple cases, you should be able to get by without using tags at all. In more complex cases, having to manage multiple clocks could quickly become burdensome. I could imagine folks ending up just keeping a `map[string]clock` to keep track of them all, which amounts to much the same thing?

Reply:
I was thinking using different variables, so leveraging the language instead of a prescribed data structure.

Reply:
The cognitive overhead of thinking about multiple clocks sounds way harder than tags.  I'm strongly against promoting it in our documentation, but there is nothing actually stopping people from doing it if tags feels that burdensome.

We could make the tag matching more explicit in our naming, at the expense of a more verbose API.  I.e.

```
trap := mClock.Trap().Now().Contains("foo")
```

rather than our present

```
trap := mClock.Trap().Now("foo")
```

It would be more extensible for additional matching strategies in future.  I personally prefer to type less in UTs and handle the "cognitive burden", but I'm open to changing it if that's an overall consensus.  

Keep in mind that this is a library for _unit testing_, and as such, the number of unique calls is typically a handful, perhaps a few dozen.  I don't think it's particularly good practice to mock time for integration tests involving more than a few components.

Tagging (labeling) systems like K8s are designed to handle clusters with tens of thousands of containers and so have a complex matching language.  Honestly for this library, just having a single string with exact matching would be sufficient for 95% of cases, but 

1. the `...string` syntax affords multiple entries, and it felt annoying to like, `panic` if you put more than one
2. there are some few use cases for multiple tags, such as components that programmatically start multiple tickers/timers, such as [apphealth](https://github.com/coder/coder/blob/7049d7a8816c7d2d2eff7769d8a718ec8aca6e17/agent/apphealth.go#L134), which then allows us to verify tickers for each app get started.